cmd/juju/commands/migrate.go

```diff
@@ -178,3 +240,162 @@ func (c *migrateCommand) getTargetControllerMacaroons() ([]macaroon.Slice, error
 	defer api.Close()
 	return httpbakery.MacaroonsForURL(jar, api.CookieURL()), nil
 }
+
+func (c *migrateCommand) checkMigrationFeasibility(spec *controller.MigrationSpec) error {
+	var (
+		srcControllerName, srcModelName string
+		srcUsers, dstUsers              set.Strings
+		err                             error
+	)
+
+	if srcControllerName, err = c.ControllerName(); err != nil {
+		return err
+	}
+	if srcModelName, err = c.ModelName(); err != nil {
+		return err
+	}
+	if srcUsers, err = c.getModelUsers(names.NewModelTag(spec.ModelUUID)); err != nil {
+		return err
+	}
+	if dstUsers, err = c.getTargetControllerUsers(); err != nil {
+		return err
+	}
+
+	// If external users have access to this model we can only allow the
+	// migration to proceed if:
+	// - the local users from src exist in the dst, and
+	// - both controllers are configured with the same identity provider URL
+	srcExtUsers := filterSet(srcUsers, func(u string) bool {
+		return strings.Contains(u, "@")
+	})
+
+	if srcExtUsers.Size() != 0 {
+		srcIdentityURL, err := c.getIdentityProviderURL(srcControllerName)
+		if err != nil {
+			return errors.Annotate(err, "looking up source controller identity provider URL")
+		}
+
+		dstIdentityURL, err := c.getIdentityProviderURL(c.targetController)
+		if err != nil {
+			return errors.Annotate(err, "looking up target controller identity provider URL")
+		}
+
+		localSrcUsers := srcUsers.Difference(srcExtUsers)
+
+		// In this case external user lookups will most likely not work.
+		// Display an appropriate error message depending on whether
+		// the local users are present in dst or not.
+		if srcIdentityURL != dstIdentityURL {
+			missing := localSrcUsers.Difference(dstUsers)
+			if missing.Size() == 0 {
+				return errors.Errorf(`cannot initiate migration of model "%s:%s" to controller %q as
+external users have been granted access to the model and the two controllers have
+different identity provider configurations. To resolve this issue you can remove
+the following list of external users from "%s:%s":
+  - %s`,
+					srcControllerName, srcModelName, c.targetController,
+					srcControllerName, srcModelName,
+					strings.Join(srcExtUsers.Values(), "\n  - "),
+				)
+			}
+
+			return errors.Errorf(`cannot initiate migration of model "%s:%s" to controller %q as
+external users have been granted access to the model and the two controllers have
+different identity provider configurations. Additionally, some of the model's local
+users do not exist in the target controller. To resolve this issue you need to remove
+the following external users from "%s:%s":
+  - %s
+
+and either migrate the following list of local users to %q or remove them from "%s:%s":
```

Comment:
There is no "migration of users", instead we should say "add the following list of users".